CMakeLists.txt

```diff
@@ -17,12 +16,73 @@ endif()
 find_package(ament_cmake REQUIRED)
 find_package(rosidl_default_generators REQUIRED)
 
+set(protobuf_MODULE_COMPATIBLE TRUE)
+find_package(Protobuf CONFIG REQUIRED)
+message(STATUS "Using protobuf ${Protobuf_VERSION}")
+
+find_package(gRPC CONFIG REQUIRED)
+message(STATUS "Using gRPC ${gRPC_VERSION}")
+
+file(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/proto)
+add_custom_command(
+  OUTPUT
+    "${CMAKE_CURRENT_BINARY_DIR}/proto/ninshiki.pb.cc"
+    "${CMAKE_CURRENT_BINARY_DIR}/proto/ninshiki.pb.h"
+    "${CMAKE_CURRENT_BINARY_DIR}/proto/ninshiki.grpc.pb.cc"
+    "${CMAKE_CURRENT_BINARY_DIR}/proto/ninshiki.grpc.pb.h"
+  COMMAND $<TARGET_FILE:protobuf::protoc>
+  ARGS
+    --grpc_out "${CMAKE_CURRENT_BINARY_DIR}/proto"
+    --cpp_out "${CMAKE_CURRENT_BINARY_DIR}/proto"
+    -I ${CMAKE_CURRENT_SOURCE_DIR}/proto
+    --plugin=protoc-gen-grpc=$<TARGET_FILE:gRPC::grpc_cpp_plugin>
+    ${CMAKE_CURRENT_SOURCE_DIR}/proto/ninshiki.proto
+  DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/proto/ninshiki.proto
+)
+
+add_library(ninshiki_proto SHARED
+  "${CMAKE_CURRENT_BINARY_DIR}/proto/ninshiki.pb.cc"
+  "${CMAKE_CURRENT_BINARY_DIR}/proto/ninshiki.pb.h"
+  "${CMAKE_CURRENT_BINARY_DIR}/proto/ninshiki.grpc.pb.cc"
+  "${CMAKE_CURRENT_BINARY_DIR}/proto/ninshiki.grpc.pb.h"
+)
+
+target_compile_options(ninshiki_proto PRIVATE -fPIC)
+target_compile_features(ninshiki_proto PUBLIC cxx_std_17)
+
+target_link_libraries(ninshiki_proto PUBLIC
+  protobuf::libprotobuf
+)
```

Comment:
It would be better to simplify this flow by creating a function and placing it in an independent gRPC utility package. This way, we could simply do this to generate a Proto library from a Proto file:

```cmake
find_package(YourGrpcUtilityPackage)

add_proto_library(ninshiki_proto proto/ninshiki.proto)
```